### PR TITLE
Refactor CreateView/EditView validation logic to support non-form validation

### PR DESCRIPTION
**Pull request by gasman**
A rabbit-hole I went down while trying to implement `Accept: application/json` error responses for the snippet create/edit views for the autosave MVP, and figured would be best handled in its own PR...

The form validation logic we inherit from Django's `BaseCreateView` / `BaseUpdateView` is heavily coupled to the `form.is_valid()` method, meaning that we need to do some pretty messy things to accommodate validation conditions that arise outside of the form:

* `wagtail.admin.views.workflows` has a formset and secondary form alongside the primary one, which are validated within `form_valid` after the primary form has been saved, relying on transactions to roll this back in case of validation error
* `CreateEditViewOptionalFeaturesMixin` overrides `post` to check the model's `locked` flag, and then `get_error_message` has to repeat that logic because `form_invalid` has no knowledge of why we've ended up there...

To fix this, we introduce a new `is_valid(form)` method on the view, which by default calls `form.is_valid()` but can be overridden to add validation checks on any other state the view is holding on to. In the case of a validation error, it sets the attribute `self.produced_error_message` (unfortunately `self.error_message` was already taken by the template string for the validation error message...) to be picked up by `form_invalid`.

### CODE REVIEW

File: wagtail/admin/views/generic/mixins.py [hunk 1]
```diff
@@ -645,6 +635,11 @@ def form_invalid(self, form):
             # Refresh the lock object as now WorkflowLock no longer applies
             self.lock = self.get_lock()
             self.locked_for_user = self.lock and self.lock.for_user(self.request.user)
+            # While this was essentially a successful action, we are repurposing the form_invalid
+            # mechanism to reshow the form. We thus need to unset produced_error_message (as set
+            # during self.is_valid()) to avoid showing an error message.
+            self.produced_error_message = None
```
Comment (laymonage) [inline wagtail/admin/views/generic/mixins.py:641]: This comment got me thinking how we ended up here – "repurposing the `form_invalid` mechanism to reshow the form" is not really done _for_ the `cancel-workflow` action. It's done for when "the object is locked for the user", to ensure the form is considered invalid even when it passes validation.

If the form is valid and the object wasn't locked for the user (e.g. when a superuser/moderator cancels a workflow they submitted), the `cancel-workflow` process is done normally via the `form_valid` mechanism (instead of repurposing `form_invalid`).

So maybe something like this is closer to what's actually happening:

```suggestion
            # While this was essentially a successful action, the form is still
            # invalid (either due to validation errors or a lock). We thus need
            # to unset produced_error_message (as set during self.is_valid()) to
            # avoid showing an error message.
```

And maybe the comment at the beginning of the method should be updated to something like

```py
        # Even if the form is invalid (either due to validation errors or a lock),
        # the original submitter can still cancel the workflow
```

## Aside

I considered refactoring this into `is_valid()`, basically something like:

```diff
diff --git a/wagtail/admin/views/generic/mixins.py b/wagtail/admin/views/generic/mixins.py
index 061670f042..3f3c8b7766 100644
--- a/wagtail/admin/views/generic/mixins.py
+++ b/wagtail/admin/views/generic/mixins.py
@@ -623,26 +623,6 @@ class CreateEditViewOptionalFeaturesMixin:

         return response

-    def form_invalid(self, form):
-        # Even if the object is locked due to not having permissions,
-        # the original submitter can still cancel the workflow
-        if self.action == "cancel-workflow":
-            self.cancel_workflow_action()
-            messages.success(
-                self.request,
-                self.get_success_message(),
-                buttons=self.get_success_buttons(),
-            )
-            # Refresh the lock object as now WorkflowLock no longer applies
-            self.lock = self.get_lock()
-            self.locked_for_user = self.lock and self.lock.for_user(self.request.user)
-            # While this was essentially a successful action, we are repurposing the form_invalid
-            # mechanism to reshow the form. We thus need to unset produced_error_message (as set
-            # during self.is_valid()) to avoid showing an error message.
-            self.produced_error_message = None
-
-        return super().form_invalid(form)
-
     def get_last_updated_info(self):
         # Create view doesn't have last updated info
         if self.view_name == "create":
@@ -783,16 +763,39 @@ class CreateEditViewOptionalFeaturesMixin:
                 _("The %(model_name)s could not be saved as it is locked")
                 % {"model_name": self.model._meta.verbose_name}
             )
-            return False
+            result = False
+            # The next logic uses elif since the form is always considered
+            # invalid if it's locked for the user, so we don't bother with
+            # further validation.

         # If saving as draft, do not enforce full validation
-        if self.saving_as_draft and isinstance(form, WagtailAdminModelForm):
+        elif self.saving_as_draft and isinstance(form, WagtailAdminModelForm):
             form.defer_required_fields()
             result = super().is_valid(form)
             form.restore_required_fields()
         else:
             result = super().is_valid(form)

+        # Even if the form is invalid (either normally or due to lock), the user
+        # should still be able to cancel the workflow. Note that in the case
+        # where the form is valid and the user is cancelling the workflow, the
+        # logic is executed normally via form_valid().
+        if not result and self.action == "cancel-workflow":
+            self.cancel_workflow_action()
+            messages.success(
+                self.request,
+                self.get_success_message(),
+                buttons=self.get_success_buttons(),
+            )
+            # Refresh the lock object as now WorkflowLock no longer applies when
+            # rendering the response
+            self.lock = self.get_lock()
+            self.locked_for_user = self.lock and self.lock.for_user(self.request.user)
+            # Remove the produced error message as the cancellation is successful
+            self.produced_error_message = None
+            # We don't change the result as we still consider the form to be invalid,
+            # whatever the original reason is, and we don't want to save it.
+
         return result


```

But then I realised we probably wouldn't want to have side-effects like database changes and adding success messages inside `is_valid()`, so I think keeping it in `form_invalid()` is better.

File: wagtail/admin/views/generic/models.py [hunk 2]
```diff
@@ -491,6 +491,31 @@ def setup(self, request, *args, **kwargs):
         super().setup(request, *args, **kwargs)
         self.action = self.get_action(request)
 
+    def post(self, request, *args, **kwargs):
+        # BaseCreateView.post() would set self.object here, but some subclasses need
+        # to do that during setup() instead, so only do that if it hasn't already been set.
+        if not hasattr(self, "object"):
+            self.object = None
```
Comment (laymonage) [inline wagtail/admin/views/generic/models.py:498]: I think eventually we should get rid of `self.object = self.get_object()` as I explained in https://github.com/wagtail/wagtail/issues/11590#issuecomment-3632351690, but for now I think this is OK

File: docs/releases/7.3.md [hunk 3]
```diff
@@ -74,3 +74,7 @@ Previously, when loading fixtures of Page models or other models extending `Tran
...
+
+### Refactored validation logic for generic model `CreateView` and `EditView`
+
+The form validation logic in `wagtail.admin.views.generic.models.CreateView` (aliased as `wagtail.admin.views.generic.CreateView`) and `wagtail.admin.views.generic.models.EditView` (aliased as `wagtail.admin.views.generic.EditView`) has been refactored to better support validation conditions not handled by the form object (such as formsets displayed alongside the form, or a 'locked' status on the model). Any user code that overrides the `post`, `form_valid`, `form_invalid` and/or `get_error_message` methods to implement additional validation will need to be updated. These views now implement a new method `is_valid(form)`, which by default delegates to `form.is_valid()`. To signal a validation error, this method should set the attribute `self.produced_error_message` to the desired error message string, and return `False`. For a generic validation error, this error message string can be obtained from `self.get_error_message()`.
```
Comment (laymonage) [inline docs/releases/7.3.md:80]: Should we add a hint that says this applies whether you use the views directly/via `ModelViewSet`/via `SnippetViewSet`? Otherwise at a glance people might not think this is relevant for them (particularly for snippets, as it might not be immediately clear that it uses subclasses of the generic views).

File: wagtail/admin/views/generic/mixins.py (hunk 1 above)
Comment (gasman) [reply wagtail/admin/views/generic/mixins.py:641]: My first attempt at this was to add a helper property `is_cancelling_workflow_on_locked_object` and treat that as a 'valid' condition as a special case that short-circuits the rest of the logic in `is_valid` and `form_valid` - but then that would have meant either proceeding with the standard "redirect away" behaviour for successful submissions or reconstructing the "reshow form" response within form_valid, and having to make an explicit decision about the preferred UX for this action (rather than just keeping the existing code path) felt like a yak-shave too far :-)

I think this comment was written from that perspective - I'm regarding this as a successful "cancel workflow on a locked object" operation that happens to be processed through `form_invalid` as an implementation quirk (that therefore needs to be called out in a code comment), rather than a failed form submission that has a side effect of cancelling the workflow.

File: wagtail/admin/views/generic/models.py (hunk 2 above)
Comment (gasman) [reply wagtail/admin/views/generic/models.py:498]: Yep, this feels like an ideal case for a cached property, but given how much of the existing code seems to be sensitive to exactly where `self.object` becomes available - for example, the changes to `wagtail/admin/forms/workflows.py` in this PR that were necessary because the old code was implicitly relying on receiving a `None` rather than an unsaved Workflow instance - this would likely require a coordinated change across all subclasses, which is more than I wanted to get into here...

File: wagtail/admin/views/generic/mixins.py (hunk 1 above)
Comment (gasman) [reply wagtail/admin/views/generic/mixins.py:641]: Ah, I see your point now - the original comment saying

        # Even if the object is locked due to not having permissions,
        # the original submitter can still cancel the workflow

led me to believe that we only ever took this path when the object was locked, but in fact it is possible for an admin (or other user with a moderator role on the workflow) to submit for moderation (leaving it unlocked for them), introduce a validation error, and cancel. ~~The new code arguably handles this better, since it doesn't suppress the legitimate validation error :-)~~ (actually, yes it does - not sure why i thought otherwise)

I don't think it'll be _too_ onerous to write a test for this, so I'll do that for completeness.

Comment (gasman) [reply wagtail/admin/views/generic/mixins.py:641]: On further inspection I think this is adequately tested by e.g. https://github.com/wagtail/wagtail/blob/433cc7877f38aa06592632cf05368b532c972df1/wagtail/admin/tests/test_workflows.py#L2840

File: docs/releases/7.3.md (hunk 3 above)
Comment (gasman) [reply docs/releases/7.3.md:80]: I don't think we need to bring viewsets into the conversation, because as I see it developers will only be affected by this if they've specifically customised the form handling logic within `CreateView` / `EditView`, and in that case the advice is the same whether viewsets are involved or not. We should indeed mention that this applies to the `wagtail.snippets.views` CBVs, though.

**Comment by gasman** [comment]
Merged (with suggested comment / documentation updates) in 4873aea023aab96a2e76144e9589eae574db6c4a.

File: wagtail/admin/views/generic/mixins.py (hunk 1 above)
Comment (laymonage) [reply wagtail/admin/views/generic/mixins.py:641]: > in fact it is possible for an admin (or other user with a moderator role on the workflow) to submit for moderation (leaving it unlocked for them), introduce a validation error, and cancel.

Exactly, yep. That's why I suggested the comment at the start of the method to be updated as well – I think we can all benefit from a comment that explains that edge case.

> On further inspection I think this is adequately tested by e.g.

It took me a bit of time to figure out why that test was failing when I was playing around with the code to come up with the suggestion, but yes, that test submits an invalid form – so I think we're covered. Thanks!
